File: sdks/java/src/main/java/com/backlex/BacklexClient.java
package com.backlex;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

/**
 * The official Java client for the backlex API — a thin, typed wrapper over the
 * same REST + SSE surface the TypeScript SDK (@backlex/client) speaks. Build with
 * {@link #builder(String)}. Three auth modes: server key, workspace app mode
 * (token capture), or cookie session. Calls are synchronous and throw the
 * unchecked {@link BacklexException} on failure.
 */
public final class BacklexClient {

    static final ObjectMapper MAPPER = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
            .setVisibility(PropertyAccessor.GETTER, JsonAutoDetect.Visibility.NONE)
            .setVisibility(PropertyAccessor.IS_GETTER, JsonAutoDetect.Visibility.NONE)
            .setSerializationInclusion(JsonInclude.Include.NON_NULL);

    final String url;
    final String apiKey;
    final String workspace;
    final String tenant;
    volatile String appToken;
    final HttpClient http;

    public final Auth auth;
    public final Storage storage;

    private BacklexClient(Builder b) {
        this.url = b.url.endsWith("/") ? b.url.substring(0, b.url.length() - 1) : b.url;
        this.apiKey = b.apiKey;
        this.workspace = b.workspace;
        this.tenant = b.tenant;
        this.appToken = b.token;
        // A cookie manager keeps same-origin cookie sessions working across calls.
        this.http = b.http != null
                ? b.http
                : HttpClient.newBuilder()
                    .cookieHandler(new java.net.CookieManager())
                    .followRedirects(HttpClient.Redirect.NORMAL)
                    .build();
        this.auth = new Auth(this);
        this.storage = new Storage(this);
    }

    public static Builder builder(String baseUrl) {
        return new Builder(baseUrl);
    }

    /** Fluent options for {@link BacklexClient}. */
    public static final class Builder {
        private final String url;
        private String apiKey;
        private String workspace;
        private String token;
        private String tenant;
        private HttpClient http;

        private Builder(String url) {
            this.url = url;
        }

        /** Static server key (pak_...) sent as a bearer on every call. */
        public Builder apiKey(String key) { this.apiKey = key; return this; }

        /** Workspace slug — puts the client in app mode (auth + token capture). */
        public Builder workspace(String slug) { this.workspace = slug; return this; }

        /** Restore a previously-saved workspace session token (app mode). */
        public Builder token(String token) { this.token = token; return this; }

        /** Scope every request to a tenant/workspace (slug or id) via the
         *  X-Backlex-Tenant header — for anonymous public reads or a pak_ key
         *  addressing a tenant other than its home one. */
        public Builder tenant(String tenant) { this.tenant = tenant; return this; }

        /** Custom HttpClient (timeouts, proxies, testing). */
        public Builder httpClient(HttpClient http) { this.http = http; return this; }

        public BacklexClient build() {
            return new BacklexClient(this);
        }
    }

    void applyAuth(HttpRequest.Builder rb) {
        if (apiKey != null && !apiKey.isEmpty()) {
            rb.header("Authorization", "Bearer " + apiKey);
        } else if (appToken != null && !appToken.isEmpty()) {
            rb.header("Authorization", "Bearer " + appToken);
        }
        if (tenant != null && !tenant.isEmpty()) {
            rb.header("X-Backlex-Tenant", tenant);
        }
    }

    /** Typed CRUD handle for a collection. */
    public <T> Collection<T> from(String slug, Class<T> type) {
        return new Collection<>(this, slug, type);
    }

    /** Raw escape hatch — issues a request with auth headers applied. */
    @SuppressWarnings("unchecked")
    <R> R request(String method, String path, Object body, JavaType type) {
        String payload = null;
        if (body != null) {
            try {
                payload = MAPPER.writeValueAsString(body);
            } catch (Exception e) {
                throw new BacklexException(0, "ENCODE", e.getMessage(), null);
            }
        }
        HttpRequest.Builder rb = HttpRequest.newBuilder(URI.create(url + path))
                .method(method, payload == null
                        ? HttpRequest.BodyPublishers.noBody()
                        : HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8))
                .header("Content-Type", "application/json");
        applyAuth(rb);

        HttpResponse<String> resp;
        try {
            resp = http.send(rb.build(), HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new BacklexException(0, "NETWORK", e.getMessage(), null);
        }

        int sc = resp.statusCode();
        if (sc < 200 || sc >= 300) {
            throw makeError(sc, resp.body());
        }
        String text = resp.body();
        if (sc == 204 || text == null || text.isEmpty() || type == null) {
            return null;
        }
        try {
            return (R) MAPPER.readValue(text, type);
        } catch (Exception e) {
            throw new BacklexException(sc, "PARSE", e.getMessage(), null);
        }
    }

    BacklexException makeError(int status, String body) {
        String code = "UNKNOWN";
        String message = "HTTP " + status;
        Object details = null;
        if (body != null && !body.isEmpty()) {
            try {
                JsonNode root = MAPPER.readTree(body);
                JsonNode err = root.get("error");
                if (err != null && err.isObject()) {
                    if (err.hasNonNull("code")) code = err.get("code").asText();
                    if (err.hasNonNull("message")) message = err.get("message").asText();
                    if (err.has("details")) details = MAPPER.convertValue(err.get("details"), Object.class);
                }
            } catch (Exception ignored) {
                // Non-JSON error body — keep the generic message.
            }
        }
        return new BacklexException(status, code, message, details);
    }

    /**
     * Serialize a ListQuery into a URL query string (mirrors buildSearch in
     * index.ts). The filter is compact JSON, percent-encoded exactly once.
     */
    static String buildSearch(ListQuery q) {
        if (q == null) {
            return "";
        }
        List<String> parts = new ArrayList<>();
        if (q.filter != null && !q.filter.isEmpty()) {
            try {
                parts.add("filter=" + enc(MAPPER.writeValueAsString(q.filter)));
            } catch (Exception e) {
                throw new BacklexException(0, "ENCODE", e.getMessage(), null);
            }
        }
        if (!q.sort.isEmpty()) {
            parts.add("sort=" + enc(String.join(",", q.sort)));
        }
        if (!q.fields.isEmpty()) {
            parts.add("fields=" + enc(String.join(",", q.fields)));
        }
        if (q.limit != null) {
            parts.add("limit=" + q.limit);
        }
        if (q.offset != null) {
            parts.add("offset=" + q.offset);
        }
        if (q.meta != null && !q.meta.isEmpty()) {
            parts.add("meta=" + enc(q.meta));
        }
        return parts.isEmpty() ? "" : "?" + String.join("&", parts);
    }

    private static String enc(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    // -- Realtime (SSE) ------------------------------------------------------

    /**
     * Subscribe to a realtime channel (e.g. "items:posts"). Returns a
     * {@link Subscription}; {@link Subscription#close()} unsubscribes. The reader
     * runs on a daemon thread and auto-reconnects on a dropped stream (3s
     * back-off), replaying via Last-Event-ID. {@code onError} may be null.
     */
    public <T> Subscription subscribe(
            String channel,
            Class<T> type,
            Consumer<ItemEvent<T>> onEvent,
            Consumer<Throwable> onError) {
        Subscription sub = new Subscription();
        JavaType evType = MAPPER.getTypeFactory().constructParametricType(ItemEvent.class, type);
        String surl = url + "/api/realtime/" + channel + "/subscribe";
        Thread t = new Thread(() -> runSse(surl, sub, evType, onEvent, onError), "backlex-sse:" + channel);
        t.setDaemon(true);
        sub.attachThread(t);
        t.start();
        return sub;
    }

    private <T> void runSse(
            String surl,
            Subscription sub,
            JavaType evType,
            Consumer<ItemEvent<T>> onEvent,
            Consumer<Throwable> onError) {
        String lastId = null;
        while (!sub.isStopped()) {
            try {
                HttpRequest.Builder rb = HttpRequest.newBuilder(URI.create(surl))
                        .GET()
                        .header("Accept", "text/event-stream");
                applyAuth(rb);
                if (lastId != null) {
                    rb.header("Last-Event-ID", lastId);
                }
                HttpResponse<InputStream> resp = http.send(rb.build(), HttpResponse.BodyHandlers.ofInputStream());
                if (resp.statusCode() != 200) {
                    if (onError != null) {
                        onError.accept(new BacklexException(resp.statusCode(), "UNKNOWN", "HTTP " + resp.statusCode(), null));
                    }
                } else {
                    sub.setStream(resp.body());
                    lastId = readSse(resp.body(), sub, evType, onEvent, onError, lastId);
                }
            } catch (Exception e) {
                if (!sub.isStopped() && onError != null) {
                    onError.accept(e);
                }
            }
            if (sub.isStopped()) {
                return;
            }
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                return;
            }
        }
    }

    @SuppressWarnings("unchecked")
    private <T> String readSse(
            InputStream in,
            Subscription sub,
            JavaType evType,
            Consumer<ItemEvent<T>> onEvent,
            Consumer<Throwable> onError,
            String lastId) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
        List<String> data = new ArrayList<>();
        String line;
        while (!sub.isStopped() && (line = br.readLine()) != null) {
            if (line.isEmpty()) {
                if (!data.isEmpty()) {
                    String payload = String.join("\n", data);
                    data.clear();
                    try {
                        ItemEvent<T> ev = (ItemEvent<T>) MAPPER.readValue(payload, evType);
                        if (ev != null) {
                            onEvent.accept(ev);
                        }
                    } catch (Exception ex) {
                        if (onError != null) {
                            onError.accept(ex);
                        }
                    }
                }
            } else if (line.charAt(0) == ':') {
                // Comment / heartbeat frame.
            } else if (line.startsWith("id:")) {
                lastId = line.substring(3).trim();
            } else if (line.startsWith("data:")) {
                String d = line.substring(5);
                if (d.startsWith(" ")) {
                    d = d.substring(1);
                }
                data.add(d);
            }
        }
        return lastId;
    }
}

File: sdks/kotlin/src/main/kotlin/com/backlex/BacklexClient.kt
package com.backlex

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.net.CookieManager
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets

/**
 * The official Kotlin client for the backlex API — a thin, typed wrapper over the
 * same REST + SSE surface the TypeScript SDK (@backlex/client) speaks. Build with
 * [BacklexClient.builder]. Three auth modes: server key, workspace app mode
 * (token capture), or cookie session. Calls are synchronous and throw the
 * unchecked [BacklexException] on failure.
 */
class BacklexClient internal constructor(
    private val url: String,
    private val apiKey: String?,
    internal val workspace: String?,
    token: String?,
    private val tenant: String?,
    private val http: HttpClient,
) {
    internal val mapper: ObjectMapper = ObjectMapper()
        .registerKotlinModule()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        .setSerializationInclusion(JsonInclude.Include.NON_NULL)

    @Volatile
    internal var appToken: String? = token

    val auth: Auth = Auth(this)
    val storage: Storage = Storage(this)

    /** Typed CRUD handle for a collection (reified convenience). */
    inline fun <reified T> from(slug: String): Collection<T> = from(slug, T::class.java)

    /** Typed CRUD handle for a collection. */
    fun <T> from(slug: String, type: Class<T>): Collection<T> =
        Collection(this, slug, mapper.typeFactory.constructType(type))

    internal fun applyAuth(rb: HttpRequest.Builder) {
        when {
            !apiKey.isNullOrEmpty() -> rb.header("Authorization", "Bearer $apiKey")
            !appToken.isNullOrEmpty() -> rb.header("Authorization", "Bearer $appToken")
        }
        if (!tenant.isNullOrEmpty()) rb.header("X-Backlex-Tenant", tenant)
    }

    /** Raw escape hatch — issues a request with auth headers applied. */
    @Suppress("UNCHECKED_CAST")
    fun <R> request(method: String, path: String, body: Any?, type: JavaType?): R {
        val payload = body?.let { mapper.writeValueAsString(it) }
        val rb = HttpRequest.newBuilder(URI.create(url + path))
            .method(
                method,
                if (payload == null) HttpRequest.BodyPublishers.noBody()
                else HttpRequest.BodyPublishers.ofString(payload),
            )
            .header("Content-Type", "application/json")
        applyAuth(rb)

        val resp = try {
            http.send(rb.build(), HttpResponse.BodyHandlers.ofString())
        } catch (e: Exception) {
            throw BacklexException(0, "NETWORK", e.message ?: "network error")
        }

        val sc = resp.statusCode()
        if (sc < 200 || sc >= 300) throw makeError(sc, resp.body())
        val text = resp.body()
        if (sc == 204 || text.isNullOrEmpty() || type == null) return null as R
        return try {
            mapper.readValue<Any>(text, type) as R
        } catch (e: Exception) {
            throw BacklexException(sc, "PARSE", e.message ?: "parse error")
        }
    }

    /** Raw-body request (e.g. storage uploads) with a custom content type. */
    @Suppress("UNCHECKED_CAST")
    internal fun <R> sendRaw(method: String, path: String, body: ByteArray, contentType: String?, type: JavaType?): R {
        val rb = HttpRequest.newBuilder(URI.create(url + path))
            .method(method, HttpRequest.BodyPublishers.ofByteArray(body))
        if (!contentType.isNullOrEmpty()) rb.header("Content-Type", contentType)
        applyAuth(rb)
        val resp = try {
            http.send(rb.build(), HttpResponse.BodyHandlers.ofString())
        } catch (e: Exception) {
            throw BacklexException(0, "NETWORK", e.message ?: "network error")
        }
        val sc = resp.statusCode()
        if (sc < 200 || sc >= 300) throw makeError(sc, resp.body())
        val text = resp.body()
        if (text.isNullOrEmpty() || type == null) return null as R
        return mapper.readValue<Any>(text, type) as R
    }

    /** Raw byte download. */
    internal fun downloadRaw(path: String): ByteArray {
        val rb = HttpRequest.newBuilder(URI.create(url + path)).GET()
        applyAuth(rb)
        val resp = try {
            http.send(rb.build(), HttpResponse.BodyHandlers.ofByteArray())
        } catch (e: Exception) {
            throw BacklexException(0, "NETWORK", e.message ?: "network error")
        }
        val sc = resp.statusCode()
        if (sc < 200 || sc >= 300) throw makeError(sc, String(resp.body(), StandardCharsets.UTF_8))
        return resp.body()
    }

    private fun makeError(status: Int, body: String?): BacklexException {
        var code = "UNKNOWN"
        var message = "HTTP $status"
        var details: Any? = null
        if (!body.isNullOrEmpty()) {
            try {
                val err = mapper.readTree(body).get("error")
                if (err != null && err.isObject) {
                    err.get("code")?.takeIf { it.isTextual }?.let { code = it.asText() }
                    err.get("message")?.takeIf { it.isTextual }?.let { message = it.asText() }
                    err.get("details")?.let { details = mapper.convertValue(it, Any::class.java) }
                }
            } catch (_: Exception) {
                // Non-JSON error body — keep the generic message.
            }
        }
        return BacklexException(status, code, message, details)
    }

    /**
     * Serialize a ListQuery into a URL query string (mirrors buildSearch in
     * index.ts). The filter is compact JSON, percent-encoded exactly once.
     */
    internal fun buildSearch(q: ListQuery?): String {
        if (q == null) return ""
        val parts = mutableListOf<String>()
        q.filter?.takeIf { it.isNotEmpty() }?.let { parts.add("filter=" + enc(mapper.writeValueAsString(it))) }
        if (q.sort.isNotEmpty()) parts.add("sort=" + enc(q.sort.joinToString(",")))
        if (q.fields.isNotEmpty()) parts.add("fields=" + enc(q.fields.joinToString(",")))
        if (q.expand.isNotEmpty()) parts.add("expand=" + enc(q.expand.joinToString(",")))
        q.limit?.let { parts.add("limit=$it") }
        q.offset?.let { parts.add("offset=$it") }
        q.meta?.let { parts.add("meta=" + enc(it)) }
        q.locale?.let { parts.add("locale=" + enc(it)) }
        q.q?.let { parts.add("q=" + enc(it)) }
        return if (parts.isEmpty()) "" else "?" + parts.joinToString("&")
    }

    private fun enc(s: String): String = URLEncoder.encode(s, StandardCharsets.UTF_8)

    // -- Realtime (SSE) ------------------------------------------------------

    /** Subscribe to a realtime channel (reified convenience). */
    inline fun <reified T> subscribe(
        channel: String,
        noinline onEvent: (ItemEvent<T>) -> Unit,
        noinline onError: ((Throwable) -> Unit)? = null,
    ): Subscription = subscribe(channel, T::class.java, onEvent, onError)

    /**
     * Subscribe to a realtime channel (e.g. "items:posts"). Returns a
     * [Subscription]; [Subscription.close] unsubscribes. The reader runs on a
     * daemon thread and auto-reconnects on a dropped stream (3s back-off),
     * replaying via Last-Event-ID. [onError] may be null.
     */
    fun <T> subscribe(
        channel: String,
        type: Class<T>,
        onEvent: (ItemEvent<T>) -> Unit,
        onError: ((Throwable) -> Unit)? = null,
    ): Subscription {
        val sub = Subscription()
        val evType = mapper.typeFactory.constructParametricType(ItemEvent::class.java, type)
        val surl = "$url/api/realtime/$channel/subscribe"
        val thread = Thread({ runSse(surl, sub, evType, onEvent, onError) }, "backlex-sse:$channel")
        thread.isDaemon = true
        sub.attachThread(thread)
        thread.start()
        return sub
    }

    private fun <T> runSse(
        surl: String,
        sub: Subscription,
        evType: JavaType,
        onEvent: (ItemEvent<T>) -> Unit,
        onError: ((Throwable) -> Unit)?,
    ) {
        var lastId: String? = null
        while (!sub.stopped) {
            try {
                val rb = HttpRequest.newBuilder(URI.create(surl)).GET().header("Accept", "text/event-stream")
                applyAuth(rb)
                lastId?.let { rb.header("Last-Event-ID", it) }
                val resp = http.send(rb.build(), HttpResponse.BodyHandlers.ofInputStream())
                if (resp.statusCode() != 200) {
                    onError?.invoke(BacklexException(resp.statusCode(), "UNKNOWN", "HTTP ${resp.statusCode()}"))
                } else {
                    sub.setStream(resp.body())
                    lastId = readSse(resp.body(), sub, evType, onEvent, onError, lastId)
                }
            } catch (e: Exception) {
                if (!sub.stopped) onError?.invoke(e)
            }
            if (sub.stopped) return
            try {
                Thread.sleep(3000)
            } catch (e: InterruptedException) {
                return
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> readSse(
        input: InputStream,
        sub: Subscription,
        evType: JavaType,
        onEvent: (ItemEvent<T>) -> Unit,
        onError: ((Throwable) -> Unit)?,
        lastIdIn: String?,
    ): String? {
        var lastId = lastIdIn
        val reader = BufferedReader(InputStreamReader(input, StandardCharsets.UTF_8))
        val data = mutableListOf<String>()
        while (!sub.stopped) {
            val line = reader.readLine() ?: break
            when {
                line.isEmpty() -> {
                    if (data.isNotEmpty()) {
                        val payload = data.joinToString("\n")
                        data.clear()
                        try {
                            onEvent(mapper.readValue<Any>(payload, evType) as ItemEvent<T>)
                        } catch (e: Exception) {
                            onError?.invoke(e)
                        }
                    }
                }
                line.startsWith(":") -> {}
                line.startsWith("id:") -> lastId = line.substring(3).trim()
                line.startsWith("data:") -> {
                    var d = line.substring(5)
                    if (d.startsWith(" ")) d = d.substring(1)
                    data.add(d)
                }
            }
        }
        return lastId
    }

    class Builder(private val baseUrl: String) {
        private var apiKey: String? = null
        private var workspace: String? = null
        private var token: String? = null
        private var tenant: String? = null
        private var http: HttpClient? = null

        /** Static server key (pak_...) sent as a bearer on every call. */
        fun apiKey(key: String?) = apply { this.apiKey = key }

        /** Workspace slug — puts the client in app mode (auth + token capture). */
        fun workspace(slug: String?) = apply { this.workspace = slug }

        /** Restore a previously-saved workspace session token (app mode). */
        fun token(token: String?) = apply { this.token = token }

        /** Scope every request to a tenant/workspace (slug or id) via the
         *  X-Backlex-Tenant header. */
        fun tenant(tenant: String?) = apply { this.tenant = tenant }

        /** Custom HttpClient (timeouts, proxies, testing). */
        fun httpClient(client: HttpClient) = apply { this.http = client }

        fun build(): BacklexClient = BacklexClient(
            baseUrl.trimEnd('/'),
            apiKey,
            workspace,
            token,
            tenant,
            http ?: HttpClient.newBuilder()
                .cookieHandler(CookieManager())
                .followRedirects(HttpClient.Redirect.NORMAL)
                .build(),
        )
    }

    companion object {
        fun builder(baseUrl: String): Builder = Builder(baseUrl)
    }
}
